### PR TITLE
chore: converted `plugins/field-grid-dropdown` to TypeScript

### PR DESCRIPTION
**Pull request by btw17**
## Notes

- This is a follow-up to https://github.com/google/blockly-samples/pull/1651
- This PR replaces https://github.com/google/blockly-samples/pull/1587 and https://github.com/google/blockly-samples/pull/1588

### CODE REVIEW

File: plugins/field-grid-dropdown/tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+        "declaration": true,
+        "declarationMap": true,
+        "module": "ES2015",
+        "moduleResolution": "node",
+        "target": "ES2015",
+        "strict": true,
+    },
+    // NOTE: `test/**/*` is automatically included in `blockly-scripts start`.
+    // Only src matters for production builds.
+    "exclude": ["test/**/*", "dist/**/*", "build/**/*"]
```
Comment (BeksOmega) [inline plugins/field-grid-dropdown/tsconfig.json:14]: Similar question about excludes as in previous PR: https://github.com/google/blockly-samples/pull/1651#discussion_r1154982424

File: plugins/field-grid-dropdown/src/index.ts [hunk 2]
```diff
@@ -9,169 +9,215 @@
  * @author kozbial@google.com (Monica Kozbial)
  */
 
-import Blockly from 'blockly/core';
+import * as Blockly from 'blockly/core';
+
+/**
+ * A config object for defining a field grid dropdown.
+ */
+export interface FieldGridDropdownConfig extends Blockly.FieldConfig {
```
Comment (BeksOmega) [inline plugins/field-grid-dropdown/src/index.ts]: I /think/ this should extend the `FieldDropdownConfig` instead, because we are passing the config to `super`. So if we ever define more config options for the `FieldDropdown` those will be supported by the `FieldGridDropdown`

File: plugins/field-grid-dropdown/src/index.ts [hunk 3]
```diff
@@ -9,169 +9,215 @@
...
-   * Sets the number of columns on the grid. Updates the styling to reflect.
-   * @param {number} columns The number of columns. Is rounded to
-   *    an integer value and must be greater than 0. Invalid
-   *    values are ignored.
-   * @private
-   */
-  setColumns(columns) {
-    this.setColumnsInternal_(columns);
-    this.updateColumnsStyling_();
+    return new FieldGridDropdown(config.options, undefined, config);
```
Comment (BeksOmega) [inline plugins/field-grid-dropdown/src/index.ts]: As per comment, it looks like this should be doing `new this`. Is that possible to maintain?

File: plugins/field-grid-dropdown/src/index.ts [hunk 4]
```diff
@@ -9,169 +9,215 @@
...
-  }
-
-  /**
-   * Sets the number of columns on the grid. Updates the styling to reflect.
-   * @param {number} columns The number of columns. Is rounded to
-   *    an integer value and must be greater than 0. Invalid
-   *    values are ignored.
-   * @private
-   */
-  setColumns(columns) {
```
Comment (BeksOmega) [inline plugins/field-grid-dropdown/src/index.ts]: Why was this removed?

File: plugins/field-grid-dropdown/src/index.ts [hunk 5]
```diff
@@ -9,169 +9,215 @@
...
+   *  must be greater than 0. Invalid values are ignored.
    */
-  setColumnsInternal_(columns) {
-    columns = parseInt(String(columns));
-    if (!isNaN(columns) && columns >= 1) {
-      this.columns_ = columns;
+  private setColumnsInternal_(columns: string | number) {
+    const cols = typeof columns === 'string' ? parseInt(columns) : columns;
+    if (!isNaN(cols) && cols >= 1) {
+      this.columns = cols;
```
Comment (BeksOmega) [inline plugins/field-grid-dropdown/src/index.ts:122]: Yay goodbye underscore!

File: plugins/field-grid-dropdown/src/index.ts [hunk 6]
```diff
@@ -9,169 +9,215 @@
...
-   * @private
+   *
+   * @param columns The number of columns. Is rounded to an integer value and
+   *  must be greater than 0. Invalid values are ignored.
    */
-  setColumnsInternal_(columns) {
-    columns = parseInt(String(columns));
-    if (!isNaN(columns) && columns >= 1) {
-      this.columns_ = columns;
+  private setColumnsInternal_(columns: string | number) {
```
Comment (BeksOmega) [inline plugins/field-grid-dropdown/src/index.ts]: The underscore can actually be removed here as well (since this method is private).

Here and for all other private methods (or it can be left for a follow-up)

File: plugins/field-grid-dropdown/src/index.ts [hunk 7]
```diff
@@ -9,169 +9,215 @@
...
-        this.menu_.getElement(), 'fieldGridDropDownContainer');
+    const menuElement = this.menu_?.getElement() ?? null;
+    if (menuElement) {
+      Blockly.utils.dom.addClass(menuElement,
+          'fieldGridDropDownContainer');
+    }
     this.updateColumnsStyling_();
 
+    // NOTE: dropdownDispose is 'private' in FieldDropdown. It is corrected
+    // in a later version to be 'protected', though until that is
```
Comment (BeksOmega) [inline plugins/field-grid-dropdown/src/index.ts]: I think this is fixed in v9.3.0!

File: plugins/field-grid-dropdown/src/index.ts [hunk 8]
```diff
@@ -9,169 +9,215 @@
...
+    return {
+      primary: this.primaryColour ?? colourSource.getColour(),
+      border: this.borderColour ?? colourSource.getColourTertiary(),
+    };
+  }
 }
 
+// IRegisterableField should accept `T extends Field`.
+// This `fromJson` doesn't match, though it should be compatible. The options
+// rightfully should be required here.
```
Comment (BeksOmega) [inline plugins/field-grid-dropdown/src/index.ts]: Is this comment still relevant?

**Review by BeksOmega — COMMENTED** [review]
Generally LGTM, just a few questions =)

File: plugins/field-grid-dropdown/tsconfig.json (hunk 1 above)
Comment (btw17) [reply plugins/field-grid-dropdown/tsconfig.json:14]: Responded in the previous PR!

Tldr; this is a mostly innocuous change that just makes VS Code complain less when the TS Config file is open with artifacts built. It doesn't actually impact the pipeline build since those artifacts aren't included in git anyway.

File: plugins/field-grid-dropdown/src/index.ts (hunk 2 above)
Comment (btw17) [reply plugins/field-grid-dropdown/src/index.ts]: Good catch!

File: plugins/field-grid-dropdown/src/index.ts (hunk 8 above)
Comment (btw17) [reply plugins/field-grid-dropdown/src/index.ts]: Looks like it isn't - updated!

File: plugins/field-grid-dropdown/src/index.ts (hunk 7 above)
Comment (btw17) [reply plugins/field-grid-dropdown/src/index.ts]: Updated!

File: plugins/field-grid-dropdown/src/index.ts (hunk 6 above)
Comment (btw17) [reply plugins/field-grid-dropdown/src/index.ts]: Done!

File: plugins/field-grid-dropdown/src/index.ts (hunk 3 above)
Comment (btw17) [reply plugins/field-grid-dropdown/src/index.ts]: Looks like `this` works fine. Updated!

File: plugins/field-grid-dropdown/src/index.ts (hunk 4 above)
Comment (btw17) [reply plugins/field-grid-dropdown/src/index.ts]: I'll need to check back on this later this week when I've got time. Some of the changes were pulled from previous work I'd done before the `9.3.1` updates, so if this plugin changed in the meantime it's possible I missed something.

Comment (btw17) [reply plugins/field-grid-dropdown/src/index.ts]: I see now - it was marked `private`, though it's not used anywhere internally. There's no purpose in it being here if it's private as it's currently not used and can't be used by subclasses.

I could add it back without marking it `private`, or I could leave it as is. WDYT?

Comment (BeksOmega) [reply plugins/field-grid-dropdown/src/index.ts]: If it works (i.e. properly rerenders the dropdown div) I think it makes sense to readd it as public! Seems like a reasonable part of the API this field should have =)

But if it doesn't actually work properly (and can't trivially be fixed). Then I'd say leave as is (i.e. removed). That SGTY?

Comment (btw17) [reply plugins/field-grid-dropdown/src/index.ts]: SGTM! It works just fine as public, so I'll add it back.

**Comment by btw17** [comment]
All set @BeksOmega!
